source/credentials_provider_sts_web_identity.c

```diff
@@ -0,0 +1,1120 @@
+        &impl,
+        sizeof(struct aws_credentials_provider_sts_web_identity_impl));
+
+    if (!provider) {
+        return NULL;
+    }
+
+    AWS_ZERO_STRUCT(*provider);
+    AWS_ZERO_STRUCT(*impl);
+
+    aws_credentials_provider_init_base(provider, allocator, &s_aws_credentials_provider_sts_web_identity_vtable, impl);
+
+    AWS_LOGF_TRACE(AWS_LS_AUTH_CREDENTIALS_PROVIDER, "(id=%p): initializing a new tlx context", (void *)provider);
+    struct aws_tls_ctx_options tls_options;
+    aws_tls_ctx_options_init_default_client(&tls_options, allocator);
+    impl->ctx = aws_tls_client_ctx_new(allocator, &tls_options);
+
+    if (!impl->ctx) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): failed to create a tls context with error %s",
+            (void *)provider,
+            aws_error_str(aws_last_error()));
+        aws_tls_ctx_options_clean_up(&tls_options);
+        goto on_error;
+    }
+
+    aws_tls_connection_options_init_from_ctx(&impl->connection_options, impl->ctx);
+    struct aws_byte_cursor host = aws_byte_cursor_from_buf(&parameters->endpoint);
+    if (aws_tls_connection_options_set_server_name(&impl->connection_options, allocator, &host)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): failed to create a tls connection options with error %s",
+            (void *)provider,
+            aws_error_str(aws_last_error()));
+        goto on_error;
+    }
+
+    struct aws_socket_options socket_options;
+    AWS_ZERO_STRUCT(socket_options);
+    socket_options.type = AWS_SOCKET_STREAM;
+    socket_options.domain = AWS_SOCKET_IPV4;
+    socket_options.connect_timeout_ms = (uint32_t)aws_timestamp_convert(
+        STS_WEB_IDENTITY_CONNECT_TIMEOUT_DEFAULT_IN_SECONDS, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_MILLIS, NULL);
+
+    struct aws_http_connection_manager_options manager_options;
+    AWS_ZERO_STRUCT(manager_options);
+    manager_options.bootstrap = options->bootstrap;
+    manager_options.initial_window_size = STS_WEB_IDENTITY_RESPONSE_SIZE_LIMIT;
+    manager_options.socket_options = &socket_options;
+    manager_options.host = host;
+    manager_options.port = 443;
+    manager_options.max_connections = 2;
+    manager_options.shutdown_complete_callback = s_on_connection_manager_shutdown;
+    manager_options.shutdown_complete_user_data = provider;
+    manager_options.tls_connection_options = &(impl->connection_options);
+
+    impl->function_table = options->function_table;
+    if (impl->function_table == NULL) {
+        impl->function_table = &s_default_function_table;
+    }
+
+    impl->connection_manager = impl->function_table->aws_http_connection_manager_new(allocator, &manager_options);
+    if (impl->connection_manager == NULL) {
+        goto on_error;
+    }
+
+    impl->role_arn = aws_string_new_from_array(allocator, parameters->role_arn.buffer, parameters->role_arn.len);
```

Comment:
Why not just make the parameters a member?

Reply:
hmm, makes sense.